plugins/out_skywalking/skywalking.c

```diff
@@ -0,0 +1,387 @@
+#define DEFAULT_SW_SVC_NAME "sw-service"
+#define DEFAULT_SW_INS_NAME "fluent-bit"
+#define DEFAULT_SW_LOG_PATH "/v3/logs"
+
+static int cb_sw_init(struct flb_output_instance *ins, struct flb_config *config, void *data) {
+    int io_flags;
+    struct flb_output_sw *ctx;
+    const char* tmp;
+
+    /* Allocate plugin context */
+    ctx = flb_calloc(1, sizeof(struct flb_output_sw));
+    if (!ctx) {
+        flb_errno();
+        return -1;
+    }
+    ctx->ins = ins;
+
+    /* setup access token */
+    tmp = flb_output_get_property("auth_token", ins);
+    if (tmp) {
+        flb_plg_debug(ctx->ins, "OAP token: %s", ctx->auth_token);
+        ctx->auth_token = flb_sds_create(tmp);
+    }
+
+    /* setup uri */
+    ctx->uri = flb_sds_create(DEFAULT_SW_LOG_PATH);
+    if (!ctx->uri) {
+        flb_plg_error(ctx->ins, "failed to configure endpoint");
+        flb_free(ctx);
+        return -1;
+    }
+
+    /* configure service/instance name */
+    tmp = flb_output_get_property("service_name", ins);
+    if (tmp) {
+        ctx->svc_name = flb_sds_create(tmp);
+    } else {
+        ctx->svc_name = flb_sds_create(DEFAULT_SW_SVC_NAME);
+    }
+
+    tmp = flb_output_get_property("instance_name", ins);
+    if (tmp) {
+        ctx->svc_inst_name = flb_sds_create(tmp);
+    } else {
+        ctx->svc_inst_name = flb_sds_create(DEFAULT_SW_INS_NAME);
+    }
+
+    flb_plg_debug(ctx->ins, "configured %s/%s", ctx->svc_name, ctx->svc_inst_name);
+
+    flb_output_net_default(DEFAULT_SW_OAP_HOST, DEFAULT_SW_OAP_PORT, ins);
+
+    /* scheme configuration */
+    if (ins->use_tls == FLB_TRUE) {
+        io_flags = FLB_IO_TLS;
+        ctx->http_scheme = flb_sds_create("https://");
+    } else {
+        io_flags = FLB_IO_TCP;
+        ctx->http_scheme = flb_sds_create("http://");
+    }
+
+    /* setup host and port */
+    ctx->host = ins->host.name;
+    ctx->port = ins->host.port;
+    flb_plg_debug(ctx->ins, "OAP address: %s:%d", ctx->host, ctx->port);
+
+    /* configure upstream instance */
+    ctx->u = flb_upstream_create(config, ctx->host, ctx->port, io_flags, ins->tls);
+    if (!ctx->u) {
+        flb_plg_error(ctx->ins, "failed to create upstream context");
+        flb_free(ctx);
+        return -1;
+    }
+
+    flb_output_upstream_set(ctx->u, ins);
+
+    /* Set the plugin context */
+    flb_output_set_context(ins, ctx);
+    flb_output_set_http_debug_callbacks(ins);
+
+    return 0;
+}
+
+static int64_t timestamp_format(const struct flb_time* tms) {
+  int64_t timestamp = 0;
+
+  /* Format the time, use milliseconds precision not nanoseconds */
+  timestamp = tms->tm.tv_sec * 1000;
+  timestamp += tms->tm.tv_nsec / 1000000;
+
+  /* round up if necessary */
+  if (tms->tm.tv_nsec % 1000000 >= 500000) {
+      ++timestamp;
+  }
+  return timestamp;
+}
+
+static void sw_msgpack_pack_kv_str(msgpack_packer* pk, const char* key,
+                                   size_t key_len, const char *value, size_t value_len) {
```

Comment:
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions